tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/transforms/mhlo_shape_calc_placement.cc

```diff
@@ -0,0 +1,539 @@
+  assert(dict_attr && "main_func must has tf.entry_function attr");
+  auto input_placements_attr = dict_attr.get(kInputPlacementAttr);
+  if (!input_placements_attr) return PlacementType::kDevice;
+
+  SmallVector<StringRef, 4> input_placements;
+  input_placements_attr.cast<mlir::StringAttr>().getValue().split(
+      input_placements, ',', /*MaxSplit=*/-1, /*KeepEmpty=*/false);
+  assert(input_placements.size() == main_func.getNumArguments() &&
+         "input_placements.size() is not equal to num of inputs");
+  auto arg_index = hlo::getArgumentIndex(main_func, arg);
+  if (input_placements[arg_index] == hlo::kTypeHost) {
+    return PlacementType::kHost;
+  } else {
+    assert((input_placements[arg_index] == hlo::kTypeDevice) &&
+           "invalid input_placements string");
+    return PlacementType::kDevice;
+  }
+}
+
+SmallVector<llvm::StringRef, 4> PlaceShapeCalcOnHost::getOutputPlacements(
+    FuncOp main_func) {
+  auto dict_attr =
+      main_func->getAttrOfType<DictionaryAttr>("tf.entry_function");
+  assert(dict_attr && "main_func must has tf.entry_function attr");
+  auto output_placements_attr = dict_attr.get(hlo::kOutputPlacementAttr);
+  SmallVector<StringRef, 4> output_placements;
+  if (!output_placements_attr) {
+    // No placement attr is specified, thus using the inferred placement.
+    return output_placements;
+  }
+
+  output_placements_attr.cast<mlir::StringAttr>().getValue().split(
+      output_placements, ',', /*MaxSplit=*/-1, /*KeepEmpty=*/false);
+  return output_placements;
+}
+
+void PlaceShapeCalcOnHost::markI64ReturnedCPUScalarOps(
+    FuncOp func, DenseMap<Operation*, bool>& marked_ops) {
+  assert(func.getName() == "main");
+  auto return_op = func.front().getTerminator();
+  if (!isa<mlir::ReturnOp>(return_op)) return;
+
+  const auto& output_placements = getOutputPlacements(func);
+  auto returned_ops = return_op->getOperands();
+  assert(returned_ops.size() == output_placements.size());
+  for (auto output : llvm::enumerate(returned_ops)) {
+    auto idx = output.index();
+    auto op = output.value().getDefiningOp();
+    if (!op) continue;
+
+    auto dialect_name = op->getDialect()->getNamespace();
+    if (!isTargetDialect(dialect_name)) continue;
+
+    if (auto type = op->getResult(0).getType().dyn_cast<RankedTensorType>()) {
+      if ((output_placements[idx] == kTypeHost) &&
+          type.getElementType().isInteger(64) && (type.getRank() == 0)) {
+        marked_ops[op] = true;
+      }
+    }
+  }
+}
+
+void PlaceShapeCalcOnHost::markShapeCalculationOps(
+    FuncOp func, DenseMap<Operation*, bool>& marked_ops) {
+  auto& block = func.getBlocks().front();
+  block.walk([&](Operation* op) {
+    auto dialect_name = op->getDialect()->getNamespace();
+    if (!isTargetDialect(dialect_name)) return;
+    if (op->getParentOp() != func.getOperation()) return;
+
+    // 1. If the op is already marked, mark all of its operands
+    //    as shape calculation ops
```

Comment:
I don't quite get how it works? How can an operation be already marked since you walk the block in order? Users will be visited before their operands won't they?